funcx_endpoint/funcx_endpoint/executors/high_throughput/funcx_manager.py

```diff
@@ -308,6 +324,59 @@ def pull_tasks(self, kill_event):
+                        logger.warning(f"Task:{task_id} is not in task_worker_map.")
+                        logger.warning("Possible duplicate cancel or race-condition")
+                        continue
+                    # Cancel task by killing the worker it is on
+                    worker_id_raw = self.task_worker_map[task_id]['worker_id']
+                    worker_to_kill = self.task_worker_map[task_id]['worker_id'].decode('utf-8')
+                    worker_type = self.task_worker_map[task_id]['task_type']
+                    logger.debug(f"Cancelling task running on worker:{self.task_worker_map[task_id]}")
+                    try:
+                        logger.info(f"Removing worker:{worker_id_raw} from map")
+                        self.worker_map.start_remove_worker(worker_type)
+                        self.worker_map.remove_worker(worker_id_raw)
+                        logger.info(f"Popping worker:{worker_to_kill} from worker_procs")
+                        proc = self.worker_procs.pop(worker_to_kill)
+                        logger.warning(f"Sending process:{proc.pid} terminate signal")
+                        proc.terminate()
+                        try:
+                            proc.wait(1)  # Wait 1 second before attempting SIGKILL
+                        except subprocess.TimeoutExpired:
+                            logger.exception("Process did not terminate in 1 second")
+                            logger.warning(f"Sending process:{proc.pid} kill signal")
+                            proc.kill()
+                        else:
+                            logger.debug(f"Worker process exited with : {proc.returncode}")
+
+                        raise TaskCancelled(worker_to_kill, self.uid)
+                    except Exception as e:
+                        logger.exception(f"Raise exception, handling: {e}")
+                        result_package = {'task_id': task_id,
+                                          'container_id': worker_type,
+                                          'exception': self.serializer.serialize(
+                                              RemoteExceptionWrapper(*sys.exc_info()))}
+                        self.pending_result_queue.put(pickle.dumps(result_package))
+
+                    worker_proc = self.worker_map.add_worker(
+                        worker_id=str(self.worker_map.worker_id_counter),
+                        worker_type=self.worker_type,
+                        container_cmd_options=self.container_cmd_options,
+                        address=self.address,
+                        debug=self.debug,
+                        uid=self.uid,
+                        logdir=self.logdir,
+                        worker_port=self.worker_port)
+                    self.worker_procs.update(worker_proc)
+                    self.task_worker_map.pop(task_id)
+                    self.remove_task(task_id)
+                    self.task_cancel_lock.release()
```

Comment:
I'm much more comfortable with locks being used in a `with` block rather than explicitly released on only some of the code paths - eg when any of the above lines raise an exception, the lock will not be released. At the moment, I think such an exception would kill off the main Manager thread anyway, so this is a bit irrelevant right now - but if the manager thread ever gets made resilient to exceptions happening inside itself, this is probably a deadlock that will happen.